File: src/main/java/com/lightbend/model/PMMLModel.java
package com.lightbend.model;

import com.google.protobuf.Descriptors;
import org.dmg.pmml.FieldName;
import org.dmg.pmml.PMML;
import org.dmg.pmml.Visitor;
import org.jpmml.evaluator.*;
import org.jpmml.evaluator.visitors.*;
import org.jpmml.model.PMMLUtil;

import java.io.ByteArrayInputStream;
import java.util.*;

/**
 * Created by boris on 5/18/17.
 */
public class PMMLModel implements Model {

    private static List<? extends Visitor> optimizers = Arrays.asList(new ExpressionOptimizer(), new FieldOptimizer(), new PredicateOptimizer(), new GeneralRegressionModelOptimizer(), new NaiveBayesModelOptimizer(), new RegressionModelOptimizer());

    private static Map<String, String> names = createNamesMap();
    private static Map<String, String> createNamesMap() {
        Map<String, String> map = new HashMap<>();
        map.put("fixed acidity", "fixed_acidity");
        map.put("volatile acidity", "volatile_acidity");
        map.put("citric acid", "citric_acid");
        map.put("residual sugar", "residual_sugar");
        map.put("chlorides", "chlorides");
        map.put("free sulfur dioxide", "free_sulfur_dioxide");
        map.put("total sulfur dioxide", "total_sulfur_dioxide");
        map.put("density", "density");
        map.put("pH", "pH");
        map.put("sulphates", "sulphates");
        map.put("alcohol", "alcohol");
        return map;
    }

    private Evaluator evaluator;
    private FieldName tname;
    private List<InputField> inputFields;
    private Map<FieldName, FieldValue> arguments = new LinkedHashMap<>();

    public PMMLModel(byte[] input) throws Throwable{
        // unmarshal PMML
        PMML pmml = PMMLUtil.unmarshal(new ByteArrayInputStream(input));
        // Optimize model
        synchronized(this) {
            for (Visitor optimizer : optimizers) {
                optimizer.applyTo(pmml);
            }
        }

        // Create and verify evaluator
        ModelEvaluatorFactory modelEvaluatorFactory = ModelEvaluatorFactory.newInstance();
        evaluator = modelEvaluatorFactory.newModelEvaluator(pmml);
        evaluator.verify();

        // Get input/target fields
        inputFields = evaluator.getInputFields();
        TargetField target = evaluator.getTargetFields().get(0);
        tname = target.getName();
    }

    @Override
    public Object score(Object input) {
        Winerecord.WineRecord inputs = (Winerecord.WineRecord) input;
        arguments.clear();
        for(InputField field : inputFields){
           arguments.put(field.getName(), field.prepare(getValueByName(inputs,field.getName().getValue())));
        }

        // Calculate Output// Calculate Output
        Map<FieldName, ?> result = evaluator.evaluate(arguments);

        // Prepare output
        double rv = 0;
        Object tresult = result.get(tname);
        if(tresult instanceof Computable){
            String value = ((Computable)tresult).getResult().toString();
            rv = Double.parseDouble(value);
        }
        else
            rv = (Double)tresult;
        return rv;
    }

    @Override
    public void cleanup() {
        // Do nothing

    }

    // Get variable value by  name
    private double getValueByName(Winerecord.WineRecord input, String name){
        Descriptors.FieldDescriptor descriptor =  input.getDescriptorForType().findFieldByName(names.get(name));
        return (double)input.getField(descriptor);
    }
}

File: model/src/main/scala/com/lightbend/modelServer/model/PMML/PMMLModel.scala
package com.lightbend.modelServer.model.PMML

/**
 * Created by boris on 5/9/17.
 *
 * Class for PMML model
 */

import java.io.{ByteArrayInputStream, ByteArrayOutputStream}

import com.lightbend.model.modeldescriptor.ModelDescriptor
import com.lightbend.model.winerecord.WineRecord
import org.dmg.pmml.{FieldName, PMML}
import org.jpmml.evaluator.visitors._
import org.jpmml.evaluator.{Computable, FieldValue, ModelEvaluatorFactory, TargetField}
import org.jpmml.model.PMMLUtil
import com.lightbend.modelServer.model.{Model, ModelFactory, ModelToServe}

import scala.collection.JavaConversions._
import scala.collection._

class PMMLModel(inputStream: Array[Byte]) extends Model {

  var arguments = mutable.Map[FieldName, FieldValue]()

  // Marshall PMML
  val pmml = PMMLUtil.unmarshal(new ByteArrayInputStream(inputStream))

  // Optimize model// Optimize model
  PMMLModel.optimize(pmml)

  // Create and verify evaluator
  val evaluator = ModelEvaluatorFactory.newInstance.newModelEvaluator(pmml)
  evaluator.verify()

  // Get input/target fields
  val inputFields = evaluator.getInputFields
  val target: TargetField = evaluator.getTargetFields.get(0)
  val tname = target.getName

  override def score(input: AnyVal): AnyVal = {
    val inputs = input.asInstanceOf[WineRecord]
    arguments.clear()
    inputFields.foreach(field => {
      arguments.put(field.getName, field.prepare(getValueByName(inputs, field.getName.getValue)))
    })

    // Calculate Output// Calculate Output
    val result = evaluator.evaluate(arguments)

    // Prepare output
    result.get(tname) match {
      case c: Computable => c.getResult.toString.toDouble
      case v: Any => v.asInstanceOf[Double]
    }
  }

  override def cleanup(): Unit = {}

  private def getValueByName(inputs: WineRecord, name: String): Double =
    PMMLModel.names.get(name) match {
      case Some(index) => {
        val v = inputs.getFieldByNumber(index + 1)
        v.asInstanceOf[Double]
      }
      case _ => .0
    }

  override def toBytes: Array[Byte] = {
    var stream = new ByteArrayOutputStream()
    PMMLUtil.marshal(pmml, stream)
    stream.toByteArray
  }

  override def getType: Long = ModelDescriptor.ModelType.PMML.value
}

object PMMLModel extends ModelFactory {

  private val optimizers = Array(new ExpressionOptimizer, new FieldOptimizer, new PredicateOptimizer,
    new GeneralRegressionModelOptimizer, new NaiveBayesModelOptimizer, new RegressionModelOptimizer)
  def optimize(pmml: PMML) = this.synchronized {
    optimizers.foreach(opt =>
      try {
        opt.applyTo(pmml)
      } catch {
        case t: Throwable => {
          println(s"Error optimizing model for optimizer $opt")
          t.printStackTrace()
          throw t
        }
      })
  }
  private val names = Map(
    "fixed acidity" -> 0,
    "volatile acidity" -> 1, "citric acid" -> 2, "residual sugar" -> 3,
    "chlorides" -> 4, "free sulfur dioxide" -> 5, "total sulfur dioxide" -> 6,
    "density" -> 7, "pH" -> 8, "sulphates" -> 9, "alcohol" -> 10
  )

  override def create(input: ModelToServe): Model = {
      new PMMLModel(input.model)
  }

  override def restore(bytes: Array[Byte]): Model = new PMMLModel(bytes)
}